### PR TITLE
Multislice rPIE; scan position correction is still broken.

### PR DESCRIPTION
**Pull request by a4894z**
Multislice rPIE.

Scan position correction is still broken I think, but the existing code for OPR updates on quantities at slice = 0 should still be ok.

### CODE REVIEW

File: src/ptychi/reconstructors/pie.py [hunk 1]
```diff
@@ -94,46 +92,53 @@ def compute_updates(
...
-                positions.round().int() + object_.pos_origin_coords,
-                delta_o_patches[:, 0],
-                op="add",
-            )
-            # Add slice dimension.
-            delta_o = delta_o.unsqueeze(0)
-
+        
+        delta_exwv_i = (psi_prime - psi)
+        delta_o = torch.zeros_like( object_.data )
```
Comment (mdw771) [inline src/ptychi/reconstructors/pie.py:105]: Please remove extra spaces on this line and the redundant parenthesis in the line above. Don't pad function arguments or array indices with spaces.

File: src/ptychi/reconstructors/pie.py [hunk 2]
```diff
@@ -94,46 +92,53 @@ def compute_updates(
...
+                    positions.round().int() + object_.pos_origin_coords,
+                    delta_o_patches[:, 0],
+                    op="add",
+                )
+                
+                delta_o[i_slice, ... ] = delta_o_i
+                
+            delta_p_i = None
+            if probe.optimization_enabled(self.current_epoch):
+                step_weight = self.calculate_probe_step_weight( (obj_patches[ :, i_slice, ... ])[:, None, ...])
```
Comment (mdw771) [inline src/ptychi/reconstructors/pie.py:124]: Remove space padding

**Review by mdw771 — CHANGES_REQUESTED** [review]

File: src/ptychi/reconstructors/pie.py [hunk 3]
```diff
@@ -94,46 +92,53 @@ def compute_updates(
...
+                delta_p_i = step_weight * delta_exwv_i  # get delta p at each position
+                delta_p_i = self.adjoint_shift_probe_update_direction(indices, delta_p_i, first_mode_only=True)
+                
+            if i_slice > 0:
+                delta_exwv_i = self.forward_model.propagate_to_previous_slice(delta_p_i, slice_index = i_slice )
+                
         delta_pos = None
```
Comment (mdw771) [inline src/ptychi/reconstructors/pie.py:131]: You can move position correction inside the for loop and run it only when the current slice is the slice selected for position correction. You can copy the same logic from LSQML and just replace `chi` with your `delta_exwv_i`

```
...
            if self.parameter_group.probe_positions.optimization_enabled(
                self.current_epoch
            ) and i_slice == self.parameter_group.probe_positions.get_slice_for_correction(
                object_.n_slices
            ):
                mean_alpha_o = pmath.trim_mean(
                    self.alpha_object_all_pos_all_slices[indices, i_slice], 0.1, dim=0
                )
                delta_pos = self.update_probe_positions(
                    chi,
                    indices,
                    obj_patches[:, i_slice : i_slice + 1],
                    mean_alpha_o * delta_o_i,
                    self.forward_model.intermediate_variables.shifted_unique_probes[i_slice],
                )

    def update_probe_positions(self, chi, indices, obj_patches, delta_o_patches, unique_probes):
        delta_pos = self.parameter_group.probe_positions.position_correction.get_update(
            chi,
            obj_patches,
            delta_o_patches,
            unique_probes,
            self.parameter_group.object.optimizer_params["lr"],
        )
        return delta_pos
```

File: src/ptychi/reconstructors/pie.py [hunk 4]
```diff
@@ -94,46 +92,53 @@ def compute_updates(
...
+                delta_o[i_slice, ... ] = delta_o_i
+                
+            delta_p_i = None
+            if probe.optimization_enabled(self.current_epoch):
+                step_weight = self.calculate_probe_step_weight( (obj_patches[ :, i_slice, ... ])[:, None, ...])
+                delta_p_i = step_weight * delta_exwv_i  # get delta p at each position
+                delta_p_i = self.adjoint_shift_probe_update_direction(indices, delta_p_i, first_mode_only=True)
+                
+            if i_slice > 0:
+                delta_exwv_i = self.forward_model.propagate_to_previous_slice(delta_p_i, slice_index = i_slice )
```
Comment (mdw771) [inline src/ptychi/reconstructors/pie.py:129]: You might want to use the backpropagated probe update **without adjoint shift** as the `delta_exwv_i` for the next slice. In this for loop, `delta_exwv_i` is always assumed to carry the position-specific subpixel shift. `adjoint_shift_probe_update_direction` undoes this shift, so if you backpropagate the adjoint-shifted `delta_p_i`, the resulting `delta_exwv` would be **without** the position-specific shifts and would become inconsistent. In LSQML you might have noticed that `chi` is set to `delta_p_i_unshifted` at the end of each iteration of the for loop. This is where it does it. 

Also, even with the adjoint shift aside, I wonder if we should use `delta_p_i` as it is for the backpropagation. The gradient of the wavefield before modulation is computed as `delta_exwv_i * obj_slice.conj()`. On the other hand, in PIE, `delta_p_i` is calculated as `step_weight * delta_exwv_i`, and `step_weight` contains something else in addition to `obj_slice.conj()`, so `step_weight * delta_exwv_i` won't be the gradient itself. The additional computations in `step_weight` will be done again when you calculate the `delta_p_i` for the next slice, so I think here it is redundant and the input of the backprop function should be changed to `delta_exwv_i * obj_patches[:, i_slice].conj()`.

**Review by mdw771 — CHANGES_REQUESTED** [review]

File: src/ptychi/reconstructors/pie.py (hunk 1 above)
Comment (a4894z) [reply src/ptychi/reconstructors/pie.py]: done

File: src/ptychi/reconstructors/pie.py (hunk 2 above)
Comment (a4894z) [reply src/ptychi/reconstructors/pie.py]: done

File: src/ptychi/reconstructors/pie.py (hunk 3 above)
Comment (a4894z) [reply src/ptychi/reconstructors/pie.py]: I've put scan position correction into the slice loop, and have tested that both OPR and position correction (at a user defined slice) are functioning.

File: src/ptychi/reconstructors/pie.py (hunk 4 above)
Comment (a4894z) [reply src/ptychi/reconstructors/pie.py]: I've made the suggested changes.
